### PR TITLE
Fix nesting for includes

### PR DESCRIPTION
**Pull request by jlvallelonga**
fixes nested includes. query strings like this should now include data as expected:
`?include=user.company.industry.tags`

### CODE REVIEW

**Comment by jeregrine** [comment]
cc @doomspork if I could get a second set of eyes on this

**Comment by doomspork** [comment]
@jeregrine I'll take a peek today 👍

File: lib/jsonapi/serializer.ex [hunk 1]
```diff
@@ -83,7 +83,14 @@ defmodule JSONAPI.Serializer do
     if {rel_view, :include} == valid_include_view && is_data_loaded?(rel_data) do
       rel_query_includes =
         if is_list(query_includes) do
-          Keyword.get(query_includes, key, [])
+          Enum.map(query_includes, fn include ->
```
Comment (doomspork) [inline lib/jsonapi/serializer.ex]: Could we refactor this entire pipechain into a `Enum.reduce/3`?

```elixir
Enum.reduce(query_includes, [], fn 
  {_key, value}, acc -> acc ++ [value]
  _, acc -> acc
end)

```

File: lib/jsonapi/serializer.ex [hunk 2]
```diff
@@ -144,29 +152,31 @@ defmodule JSONAPI.Serializer do
...
-  defp get_includes(view, []), do: view.relationships()
-  defp get_includes(view, query_includes) when is_list(query_includes) do
-    base = view.relationships()
-    Enum.reduce(query_includes, [], &(handle_include(base, &1, &2)))
-  end
-  defp get_includes(view, include) do
-    base = view.relationships()
-    Keyword.get(base, include)
+  defp get_includes(view, query_includes) do
+    get_default_includes(view) ++ get_query_includes(view, query_includes)
```
Comment (doomspork) [inline lib/jsonapi/serializer.ex]: Let's avoid pipes for single function calls.

```elixir
includes = get_default_includes(view) ++ get_query_includes(view, query_includes)
Enum.uniq(includes)
```

File: lib/jsonapi/serializer.ex [hunk 3]
```diff
@@ -144,29 +152,31 @@ defmodule JSONAPI.Serializer do
...
-    view = Keyword.get(base, parent)
-    acc = Keyword.put(acc, parent, view)
-    handle_include(view, child, acc)
-  end
-  defp handle_include({base, :include}, child, acc) do
-    handle_include(base.relationships(), child, acc)
+  defp get_default_includes(view) do
+    rels = view.relationships()
+    default_includes = rels |> Enum.filter(fn {k, v} ->
+      case v do
```
Comment (doomspork) [inline lib/jsonapi/serializer.ex]: You don't need to use a case here, match on the function head 👍

File: lib/jsonapi/serializer.ex [hunk 4]
```diff
@@ -144,29 +152,31 @@ defmodule JSONAPI.Serializer do
...
+      end
+    end)
   end
-  defp handle_include(base, child, acc) do
-    view = if is_list(base), do: Keyword.get(base, child), else: base
-    Keyword.put(acc, child, view)
+
+  defp get_query_includes(view, query_includes) do
+    rels = view.relationships()
+    Enum.map(query_includes, fn include ->
```
Comment (doomspork) [inline lib/jsonapi/serializer.ex]: I think there is an opportunity to refactor this.  We could probably move to `Enum.reduce/3` and we can eliminate the `case` by matching on the function signature instead.

File: test/jsonapi_test.exs [hunk 5]
```diff
@@ -9,7 +9,7 @@ defmodule JSONAPITest do
     def type, do: "mytype"
     def relationships do
       [author: {JSONAPITest.UserView, :include},
-       other_user: {JSONAPITest.UserView, :include}]
+       other_user: JSONAPITest.UserView]
```
Comment (doomspork) [inline test/jsonapi_test.exs]: Will these changes result in a breaking change?  If not, let's leave the existing tests alone and add new ones to cover these changes.

Comment (jlvallelonga) [reply test/jsonapi_test.exs]: Yeah this changes the functionality. Although, I think it fixes the original intended functionality of `:include` meaning that the relationship will be included by default. I'm basing this on this bit from the readme:

```
  def relationships do
    # The post's author will be included by default
    [author: {MyApp.UserView, :include},
     comments: MyApp.CommentView]
  end
```

Thoughts?

Comment (jlvallelonga) [reply test/jsonapi_test.exs]: @doomspork @jeregrine ping

**Comment by StareIntoTheBeard** [comment]
@doomspork @jeregrine @jlvallelonga bump, any other thoughts on this one? What's needed to get this merged in?

**Comment by doomspork** [comment]
Sorry @jlvallelonga, I've been away on a business trip and didn't have too much time to go through GH.  Since these _are_ a breaking change, I'm going to have to spend some time looking through them closer.  That said, if @jeregrine wants to go forward with it 👍

**Comment by doomspork** [comment]
Thank you @jlvallelonga!  @StareIntoTheBeard I will be cutting a new release shortly 👍

**Pull request merged**